activation/validation.go

```diff
@@ -264,3 +290,140 @@ func (v *Validator) PositioningAtx(
+func withTrustedID(val types.NodeID) verifyChainOption {
+	return func(o *verifyChainOpts) {
+		o.trustedNodeID = val
+	}
+}
+
+var ErrInvalidChain = errors.New("invalid ATX chain")
+
+func verifyChain(
+	ctx context.Context,
+	db sql.Executor,
+	id, goldenATXID types.ATXID,
+	validator nipostValidator,
+	log *zap.Logger,
+	opts ...verifyChainOption,
+) error {
+	log.Info("verifying ATX chain", zap.Stringer("atx_id", id))
+	options := verifyChainOpts{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	return verifyChainWithOpts(ctx, db, id, goldenATXID, validator, log, options)
+}
+
+func verifyChainWithOpts(
+	ctx context.Context,
+	db sql.Executor,
+	id, goldenATXID types.ATXID,
+	validator nipostValidator,
+	log *zap.Logger,
+	opts verifyChainOpts,
+) error {
+	atx, err := atxs.Get(db, id)
+	if err != nil {
+		return fmt.Errorf("get atx: %w", err)
+	}
+
+	switch {
+	case atx.Validity() == types.Valid:
+		log.Debug("not verifying ATX chain", zap.Stringer("atx_id", id), zap.String("reason", "already verified"))
+		return nil
+	case atx.Validity() == types.Invalid:
+		log.Debug("not verifying ATX chain", zap.Stringer("atx_id", id), zap.String("reason", "invalid"))
+		return errors.Join(ErrInvalidChain, errors.New("atx is marked as invalid"))
+	case atx.Received().Before(opts.assumedValidTime):
+		log.Debug(
+			"not verifying ATX chain",
+			zap.Stringer("atx_id", id),
+			zap.String("reason", "assumed valid"),
+			zap.Time("received", atx.Received()),
+			zap.Time("valid_before", opts.assumedValidTime),
+		)
+		return nil
+	case atx.SmesherID == opts.trustedNodeID:
+		log.Debug("not verifying ATX chain", zap.Stringer("atx_id", id), zap.String("reason", "trusted"))
+		return nil
+	}
+
+	// validate POST fully
+	commitmentAtxId := atx.CommitmentATX
+	if commitmentAtxId == nil {
+		if atxId, err := atxs.CommitmentATX(db, atx.SmesherID); err != nil {
+			return fmt.Errorf("getting commitment atx: %w", err)
+		} else {
+			commitmentAtxId = &atxId
+		}
+	}
+	if err := validator.Post(
+		ctx,
+		atx.SmesherID,
+		*commitmentAtxId,
+		atx.NIPost.Post,
+		atx.NIPost.PostMetadata,
+		atx.NumUnits,
+	); err != nil {
+		if err := atxs.SetValidity(db, id, types.Invalid); err != nil {
+			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", id))
+		}
+		return errors.Join(ErrInvalidChain, fmt.Errorf("invalid post in ATX %s: %w", id.ShortString(), err))
+	}
+
+	err = verifyChainDeps(ctx, db, atx.ActivationTx, goldenATXID, validator, log, opts)
+	switch {
+	case err == nil:
+		if err := atxs.SetValidity(db, id, types.Valid); err != nil {
+			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", id))
+		}
+	case errors.Is(err, ErrInvalidChain):
+		if err := atxs.SetValidity(db, id, types.Invalid); err != nil {
+			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", id))
+		}
+	}
```

Comment:
I think it would be useful to make `ErrInvalidChain` a type instead of just a value with a field that contains information which ATX up the chain was invalid and mark that as invalid too:

```go
type ErrInvalidChain struct {
	ID types.ATXID
}

func (e *ErrInvalidChain) Error() string {
	return fmt.Sprintf("invalid ATX chain: %s has invalid post", e.ID.ShortString())
}
```

```suggestion
	var invalidChainErr *ErrInvalidChain
	switch {
	case err == nil:
		if err := atxs.SetValidity(db, id, types.Valid); err != nil {
			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", id))
		}
	case errors.As(err, &invalidChainErr):
		if err := atxs.SetValidity(db, id, types.Invalid); err != nil {
			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", id))
		}
		if err := atxs.SetValidity(db, invalidChainErr.ID, types.Invalid); err != nil {
			log.Warn("failed to persist atx validity", zap.Error(err), zap.Stringer("atx_id", invalidChainErr.ID))
		}
	}
```

Reply:
There is no reason to add the extra step with `atxs.SetValidity(db, invalidChainErr.ID, types.Invalid);`. The chain is automatically marked invalid up from the bad ATX because this function is recursive. It marks ATXs _invalid_ when returning up with the `ErrInvalidChain` error.

Reply:
But it only marks the latest ATX of the chain as invalid, does it not? What if Alice publishes an ATX that is stored optimistically and both Bob and Charlie use it as positioning ATX - wouldn't that cause the same invalid ATX to be validated twice when fetching Bob's and Charlies ATX and going up the chain?

Reply:
Every ATX on the chain (from the faulty ATX with invalid POST up to the top ATX) is marked as invalid as recursion "bubbles up" with the `ErrInvalidChain` error.

The faulty ATX is marked here: https://github.com/spacemeshos/go-spacemesh/blob/252ee4419065d8b0561af120efcd6064d919c436/activation/validation.go#L387-L397

All ATXs up in the chain are marked here: https://github.com/spacemeshos/go-spacemesh/blob/252ee4419065d8b0561af120efcd6064d919c436/activation/validation.go#L407-L410